### ISSUE TITLE
No response/callback from funtion "hmset"

### ISSUE THREAD

**Issue by amarkumargarg** [NodeRedis/node_redis#882]
I'm running into an intermittent error where the hmset function does not provide any response (i.e the callback is never called). I also do not get any ERROR event (client object).

Here's the code snippet:
```
var hashKey = 'key';
var redisObj = {
	subkey1: 'subkey1_value', subkey2: 'subkey2_value'
};
logger.info('Calling Redis hmset to set data on Redis.');
redisClient.hmset(hashKey, redisObj, function(err) {
	if(! err) {
		logger.info('No error while setting data on Redis. Resolving.');
		deferred.resolve();
	} else {
		logger.info('Error while setting data on Redis. Rejecting.');
		deferred.reject(err);
	}
});
```

For every failure (intermittent), I see the following behavior.
* There always a log for "Calling Redis hmset to set data on Redis."
* There's no subsequent log for either "No error while setting data on Redis. Resolving." or "Error while setting data on Redis. Rejecting."
* No error event(s) as well.

What could explain such a behavior? Is there any scenario where the hmset function will not provide a response?
Additionally, is there any way to enable internal logs for the node_redis module to see what could be happening?

I shall appreciate any thoughts or direction.

**Comment by BridgeAR** [comment]
What node_redis version and what parser do you use? I could not reproduce the issue on 2.1 with your code. What function is logger info?

Internal logging can be activated by using the debug_mode. You can pass the NODE_DEBUG=redis env to activate it or use:

```js
var redis = require('redis');
redis.debug_mode = true;
```

**Comment by amarkumargarg** [comment]
Hi Ruben,

Thanks for the reply.
This issue happens under high load conditions only.

Here are some additional details:
1. Node_redis version is 2.0.1
2. Redis server is Redis Labs (https://redislabs.com/). Our redis labs plan supports surplus memory and client connections.
3. Logger is a bunyan object. Yes, we are sure the function is present. Under normal conditions the callback is called without an error.

This issue occurs under high load conditions only. 
And only certain nodejs processes show this failure momentarily, while others work fine (during the same time)

In our scenario, there are approximately 24 independent node processes which are performing operations on redis (same function as "hmset")

We are seeing random failures, where some of thes processes stopped working as they are perpetually waiting from a callback response from the hmset function - which never comes.

Please let me know if there are any thoughts.
Meanwhile i will enable debugging.

Thanks for the help,
Amar

**Comment by BridgeAR** [comment]
Please use the hiredis parser. This should solve your issue. The js parser has a regression that will be fixed soon.

**Comment by BridgeAR** [comment]
@amarkumargarg please let me know if the issue persists with the hiredis parser.

I'll also give you a ping when the issue is resolved with the js parser :)

**Comment by BridgeAR** [comment]
Duplicate of #875

**Comment by amarkumargarg** [comment]
Thanks for the direction and feedback!

I will work on the following and get back to you with results:
1. Ensure (double check) that i am using redis_node version 2.1 
2. Switch to hire parer (hiredis parser)
3. Repeat the load test

**Comment by BridgeAR** [comment]
@amarkumargarg the issue should be fixed on the latest master.

**Comment by amarkumargarg** [comment]
Hello Ruben,

I tried running a load test with Node_Redis v2.2.1 and hiredis parser.
This time I didn't face the issue (for 2 load tests). It seems above changes fix the issue.

I shall run few more load tests to verify the same and shall report back.

Thanks for the help.

**Comment by BridgeAR** [comment]
@amarkumargarg would you be so kind and also try a load test against the javascript parser? :)

**Comment by amarkumargarg** [comment]
Sure @BridgeAR, I shall also run the load test with Javascript parser as well and shall let you know the results.

**Comment by BridgeAR** [comment]
@amarkumargarg thx a lot!

**Comment by jbergknoff** [comment]
I am experiencing pretty much the same problem as reported here: intermittent failure to get called back under high load situation with many identical processes hitting redis in parallel. I've seen the issue on `GET` and `QUIT` commands, but I'm pretty sure the specific command is irrelevant.

I am using `node_redis` version 2.3.0 and the javascript parser. As I understand it, the fix #903 landed in v2.3.0, so it does not solve the issue in my case. Also, the `GET` in my situation is against a key that does not exist, so it's unlikely to be related to response parsing (also it happens with `QUIT`). I'm not interested in using `hiredis` because a compiled module would complicated my build process too much.

Any ideas?

**Comment by BridgeAR** [comment]
@jbergknoff I refactored the js parser quite substantially after 2.3 and it would be great if you could try both: 2.4.2 and master (the later handles big lranges a lot better). Please also activate the debug mode as this would indicate where the last command was handled properly. 
Do you also have a code snipped for me to look at? I can't think of any reason why any callback could get stalled. And even though you do not want to install the hiredis parser, it would be great to use it for debugging (just use `npm install hiredis` and it'll be choosen) to check if it happens for both parsers. I guess if there's an error it might occur somewhere else than in the parser.

**Comment by BridgeAR** [comment]
Ping @jbergknoff

**Comment by jbergknoff** [comment]
@BridgeAR thanks for your response. I'll be working more on this today. I hope to give you an update in a few hours.

**Comment by jbergknoff** [comment]
Okay I'm still getting the issue on the master branch. The issue appears to be that some commands never make it to the redis server and the socket waits approximately 15 minutes before giving up (`ETIMEDOUT`). Here is debug output:

```
Send ....use1.cache.amazonaws.com:6379 id 1: *2
$7
hgetall
$25
metadata-dbdf458bf4288408

Net read ....use1.cache.amazonaws.com:6379 id 1: *6
<data>

Send ....use1.cache.amazonaws.com:6379 id 1: *4
$4
hset
$16
dbdf458bf4288408
$16
2a8fea2650970119
$8
COMPLETE

<hangs here for 10-20 minutes before ETIMEDOUT>
```

On the redis server, `CLIENT LIST` shows

```
id=75983 addr=10.0.31.210:51027 fd=23 name= age=609 idle=609 flags=N db=1 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=hgetall
```

so the `HSET` isn't being received. I think the parser is off the hook.

The connection is between an EC2 instance and an ElastiCache redis server, so the network is probably fine. The process is running in a Docker container, which could be contributing to or causing the issue, but most of the sibling processes (running in other containers) do not have issues. This happens in about 1 out of every 1000 attempts.

Any ideas?

**Comment by BridgeAR** [comment]
@jbergknoff thanks a lot for digging into this! I really appreciate it.

**Comment by lucj** [comment]
Hi, I'm encountering the same problem with version 2.8.0.
The callback of a "client.lindex" is not called anymore after something like 15 minutes, no error are raised. It's in dev, so almost no load there.

**Comment by jbergknoff** [comment]
@lucj the cause of my issue was a bug in Docker which is long since fixed. More information on my specific issue here: https://github.com/aws/amazon-ecs-agent/issues/268. It's very likely that the problem you're seeing is outside the scope of this library.

**Comment by lucj** [comment]
@jbergknoff Thanks for the link.
I'm running this stack on a Docker Swarm on AWS (Docker for AWS). It seems to have an issue then as I do not have the problem with the same stack running on a single AWS instance (not swarm one) in Docker Cloud.

**Comment by hgf2012040101160** [comment]
@lucj I have the same problem.Is there an answer please?
